Error while checking build/O2/fullCI for b9937bce006aefd1b41bc02878f3483156171600 at 2024-03-04 19:53:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
Detected critical problem in logfile digi.log
digi.log:[30484:internal-dpl-injected-dummy-sink]: [18:52:58][ERROR] Exception caught: Unable to find requested service N2o29framework15ServiceRegistryE with hash 1165381451 using salt 0 for service kind 3 
[30484:internal-dpl-injected-dummy-sink]: [18:52:58][ERROR] Exception caught: Unable to find requested service N2o29framework15ServiceRegistryE with hash 1165381451 using salt 0 for service kind 3 
[ERROR] Workflow crashed - PID 30484 (internal-dpl-injected-dummy-sink) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR]  - Device internal-dpl-injected-dummy-sink: pid 30484 (exit 128)
[ERROR] SEVERE: Device internal-dpl-injected-dummy-sink (30484) returned with 128


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/613ad6de4306a6b9657fe4da6d81e026faa81b58/slc8_x86-64/o2checkcode/1.0-local2059/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12748/b9937bce006aefd1b41bc02878f3483156171600/build_O2_fullCI/pretty.html).

--- a/Framework/Core/include/Framework/AlgorithmSpec.h
+++ b/Framework/Core/include/Framework/AlgorithmSpec.h
@@ -164,6 +164,14 @@ struct ContextElementTraits<ProcessingContext> {
   }
 };
 
+template <>
+struct ContextElementTraits<InitContext> {
+  static InitContext& get(InitContext& ctx)
+  {
+    return ctx;
+  }
+};
+
 template <StringLiteral L, typename T>
 struct ContextElementTraits<Input<L, T> const> {
   static Input<L, T> get(ProcessingContext& ctx)

--- a/Framework/Core/src/CommonDataProcessors.cxx
+++ b/Framework/Core/src/CommonDataProcessors.cxx
@@ -568,29 +568,49 @@ DataProcessorSpec CommonDataProcessors::getGlobalFairMQSink(std::vector<InputSpe
   return specifyFairMQDeviceOutputProxy("internal-dpl-injected-output-proxy", danglingOutputInputs, defaultChannelConfig.c_str());
 }
 
+void retryMetricCallback(uv_async_t* async)
+{
+  static size_t lastTimeslice = -1;
+  auto* services = (ServiceRegistryRef*)async->data;
+  auto& timesliceIndex = services->get<TimesliceIndex>();
+  auto* device = services->get<RawDeviceService>().device();
+  auto channel = device->GetChannels().find("metric-feedback");
+  auto oldestPossingTimeslice = timesliceIndex.getOldestPossibleOutput().timeslice.value;
+  if (channel == device->GetChannels().end()) {
+    return;
+  }
+  fair::mq::MessagePtr payload(device->NewMessage());
+  payload->Rebuild(&oldestPossingTimeslice, sizeof(int64_t), nullptr, nullptr);
+  auto* consumed = (size_t*)malloc(sizeof(size_t));
+  *consumed = oldestPossingTimeslice;
+
+  int64_t result = channel->second[0].Send(payload, 100);
+  // If the sending worked, we do not retry.
+  if (result != 0) {
+    // If the sending did not work, we keep trying until it actually works.
+    // This will schedule other tasks in the queue, so the processing of the
+    // data will still happen.
+    uv_async_send(async);
+  } else {
+    lastTimeslice = *consumed;
+  }
+}
+
 DataProcessorSpec CommonDataProcessors::getDummySink(std::vector<InputSpec> const& danglingOutputInputs, std::string rateLimitingChannelConfig)
 {
   return DataProcessorSpec{
     .name = "internal-dpl-injected-dummy-sink",
     .inputs = danglingOutputInputs,
-    .algorithm = AlgorithmSpec{adaptStateful([](CallbackService& callbacks) {
+    .algorithm = AlgorithmSpec{adaptStateful([](CallbackService& callbacks, DeviceState& deviceState, InitContext& ic) {
+      static uv_async_t async;
+      // The callback will only have access to the
+      async.data = new ServiceRegistryRef{ic.services()};
+      uv_async_init(deviceState.loop, &async, retryMetricCallback);
       auto domainInfoUpdated = [](ServiceRegistryRef services, size_t timeslice, ChannelIndex channelIndex) {
         LOGP(debug, "Domain info updated with timeslice {}", timeslice);
-        static size_t lastTimeslice = -1;
+        retryMetricCallback(&async);
         auto& timesliceIndex = services.get<TimesliceIndex>();
-        auto device = services.get<RawDeviceService>().device();
-        auto channel = device->GetChannels().find("metric-feedback");
         auto oldestPossingTimeslice = timesliceIndex.getOldestPossibleOutput().timeslice.value;
-        if (channel != device->GetChannels().end()) {
-          fair::mq::MessagePtr payload(device->NewMessage());
-          size_t* consumed = (size_t*)malloc(sizeof(size_t));
-          *consumed = oldestPossingTimeslice;
-          if (*consumed != lastTimeslice) {
-            payload->Rebuild(consumed, sizeof(int64_t), nullptr, nullptr);
-            channel->second[0].Send(payload);
-            lastTimeslice = *consumed;
-          }
-        }
         auto& stats = services.get<DataProcessingStats>();
         stats.updateStats({(int)ProcessingStatsId::CONSUMED_TIMEFRAMES, DataProcessingStats::Op::Set, (int64_t)oldestPossingTimeslice});
       };

